singularity-eos/eos/modifiers/ramps_eos.hpp

```diff
@@ -0,0 +1,239 @@
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_RAMPS_EOS_
+#define _SINGULARITY_EOS_EOS_RAMPS_EOS_
+
+#include "stdio.h"
+#include <cassert>
+#include <cmath>
+#include <cstdlib>
+#include <iostream>
+#include <limits>
+#include <utility>
+
+#include <ports-of-call/portability.hpp>
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+namespace singularity {
+
+using namespace eos_base;
+
+template <typename T>
+void pAlpha2SAPRampParams(const T &eos, const Real alpha0, const Real Pe, const Real Pc,
+                          Real &r0, Real &a, Real &b, Real &c) {
+  // get reference conditions
+  Real rho0, T0, sie0, P0, cv0, bmod0, dpde0, dvdt0;
+  Real rmid, r1;
+  eos.ValuesAtReferenceState(rho0, T0, sie0, P0, cv0, bmod0, dpde0, dvdt0);
+  // calculate r0
+  r0 = rho0 / alpha0;
+  // calculate rmid
+  auto rmid_func = [&](const Real x) {
+    return eos.PressureFromDensityTemperature(alpha0 * x, T0);
+  };
+  RootFinding1D::RootCounts co{};
+  // get upper bound to density informed by the reference
+  // bulk modulus
+  const Real max_exp_arg = std::log(std::numeric_limits<Real>::max() * 0.99);
+  const Real exp_arg = std::min(max_exp_arg, (2.0 * Pc - P0) / bmod0);
+  const Real rho_ub = rho0 * std::exp(exp_arg);
+  // finds where rmid_func = Pe
+  RootFinding1D::findRoot(rmid_func, Pe, rho0, r0, rho_ub, 1.e-12, 1.e-12, rmid, co);
+  // calculate r1
+  auto r1_func = [&](const Real x) { return eos.PressureFromDensityTemperature(x, T0); };
+  // finds where r1_func = Pc
+  RootFinding1D::findRoot(r1_func, Pc, rmid, r0, rho_ub, 1.e-12, 1.e-12, r1, co);
+  // a
+  a = r0 * Pe / (rmid - r0);
+  // b
+  b = r0 * (Pc - Pe) / (r1 - rmid);
+  // c
+  c = (Pc * rmid - Pe * r1) / (r0 * (Pc - Pe));
+  return;
+}
+
+template <typename T>
+class SAPRampEOS : public EosBase<SAPRampEOS<T>> {
+ public:
+  // move semantics ensures dynamic memory comes along for the ride
+  SAPRampEOS(T &&t, const Real r0, const Real a, const Real b, const Real c)
+      : t_(std::forward<T>(t)), r0_(r0), a_(a), b_(b), c_(c),
+        rmid_(r0 * (a - b * c) / (a - b)), Pmid_(a * (rmid_ / r0 - 1.0)) {
+    // add input parameter checks to ensure validity of the ramp
+    assert(r0 >= 0.0);
+    assert(a > 0.0);
+    assert(b >= 0);
+    assert(a != b);
+  }
+  SAPRampEOS() = default;
+
+  auto GetOnDevice() { return SAPRampEOS<T>(t_.GetOnDevice(), r0_, a_, b_, c_); }
+  inline void Finalize() { t_.Finalize(); }
+
+  PORTABLE_INLINE_FUNCTION
+  Real get_ramp_pressure(Real rho) const {
+    const Real p_ramp{rho < r0_     ? 0.0
+                      : rho < rmid_ ? a_ * (rho / r0_ - 1.0)
+                                    : b_ * (rho / r0_ - c_)};
```

Comment:
There are asserts in the ctor, but perhaps we should disallow r0 = 0.

Reply:
We should disallow r0=0 if we're dividing by it. But even small r0 could cause problems.

Reply:
I believe the tests will cause `r0` to be fairly small since they are ramping an ideal gas. The basic idea is that $\rho_0 = \rho(P_0, T_\mathrm{ref}) / \alpha_0$. For an ideal gas, the density will be on order 1e-03.

Reply:
I'd say that number is not small relative to machine epsilon. I am wary of judging relative orders of magnitude in a way that prevents users from running the code without a special build. I will try to add @Yurlungur's smart divides thing, but I do not want to further restrict densities besides r0 needs to be a positive and non-zero floating point value.

Reply:
now using robust divide.